src/Disks/ObjectStorages/IMetadataStorage.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <unordered_map>
+#include <Poco/Timestamp.h>
+#include <IO/ReadSettings.h>
+#include <IO/WriteSettings.h>
+#include <IO/ReadBufferFromFileBase.h>
+#include <IO/WriteBufferFromFileBase.h>
+#include <Disks/DirectoryIterator.h>
+#include <Disks/WriteMode.h>
+#include <Disks/ObjectStorages/IObjectStorage.h>
+
+namespace DB
+{
+
+struct IMetadataOperation
+{
+    virtual void execute() = 0;
+    virtual void undo() = 0;
+    virtual void finalize() {}
+    virtual ~IMetadataOperation() = default;
+};
+
+using MetadataOperationPtr = std::unique_ptr<IMetadataOperation>;
+
+struct IMetadataTransaction : private boost::noncopyable
+{
+public:
+    virtual void addOperation(MetadataOperationPtr && operation) = 0;
+    virtual void commit() = 0;
+    virtual void rollback() = 0;
+
+    virtual ~IMetadataTransaction() = default;
+};
+
+using MetadataTransactionPtr = std::shared_ptr<IMetadataTransaction>;
+
+/// Metadata storage for remote disks like DiskObjectStorage.
+/// Support some subset of Disk operations, allow to read/write only
+/// small amounts of data (strings).
+/// Tries to provide some "transactions" interface, which allow
+/// to execute operations simultaneously.
+class IMetadataStorage : private boost::noncopyable
+{
+public:
+    virtual MetadataTransactionPtr createTransaction() const = 0;
+
+    virtual const std::string & getPath() const = 0;
```

Comment:
One way to implement transactional updates of metadata might be like this: IMetadataTransaction interface provides all methods that read or update the metadata (getPath, exists, isFile, ..., moveDirectory, replaceFile), and commit() method. The transaction accumulates all the changes and tries to apply them at commit time. It also saves all accessed file and directory names and their versions. This works like optimistic locks on theses files and dirs, so that these locks are broken If another transactions meanwhile commits some changes to these files or directories.
  
The code that works with it looks like this:

```
/// Start new Tx
auto metadata_tx = storage.createTransaction();
/// Do operations with the Tx
if (!metadata_tx.exists("some_info_file"))
    return; // tx will atomatically be aborted
string info = metadata_tx.read("some_info_file");
metadata_tx.createDirectory("new_dir");
metadata_tx.copyFile("some_info_file", "new_dir/some_info_file");
/// More operations with metadata_tx
/// ...

/// Try to commit all the changes. It might fail if another transaction has committed conflicting changes 
bool success = metadata_tx.commit();

if (success)
{
   log("changes successfully committed");
   /// proceed with 
}
else
{
   log("failed to make changes to the metadata");
   /// ...
}
```

What do you think: is it feasible or am I missing some use cases?